lib/jobs/emc-compose-system.js

```diff
@@ -0,0 +1,206 @@
+    }
+    
+    util.inherits(EmcComposeSystemJob, BaseJob);
+
+    /**
+     * @memberOf EmcComposeSystemJob
+     */
+    EmcComposeSystemJob.prototype._run = function() {
+        var self = this;
+        return self.composeSystem()
+        .then(function() {
+            self._done();
+        })
+        .catch(function(err) {
+            self._done(err);
+        });
+    };
+    
+    /**
+     * @function setAllocatedState
+     * @description Mark allocated state for provisioned endpoints
+     */
+    EmcComposeSystemJob.prototype.setAllocatedState = function(endpoints, id, state) {
+        var self = this;
+        return waterline.catalogs.findMostRecent({
+            node: self.nodeId,
+            source: 'Elements' 
+        })
+        .then(function(catalog) {
+            if (catalog) {
+                _.forEach(catalog.data, function(element) {
+                    var elementName = element.Type + element.Id;
+                    var match = _.first(_.filter(endpoints, function(endpoint) {
+                        return endpoint === elementName;
+                    }));
+                    if (match) {
+                        element.Allocated = { 
+                            value: state, 
+                            systemId: id
+                        };
+                    }
+                });
+                return waterline.catalogs.updateOne(
+                    { id: catalog.id },
+                    { data: catalog.data }
+                );
+            }
+        });    
+    };
+    
+    /**
+     * @function getAllocatedState
+     * @description Return current endpoints for systemId that match 
+     *              the specified state
+     */
+    EmcComposeSystemJob.prototype.getAllocatedState = function(id, state) {
+        var self = this;
+        return waterline.catalogs.findMostRecent({
+            node: self.nodeId,
+            source: 'Elements' 
+        })
+        .then(function(catalog) {
+            if (catalog) {
+                var elements = _.filter(catalog.data, function(element) {
+                    if (element.Allocated.systemId === id) {
+                        return element.Allocated.value === state;
+                    }
+                });
+                var endpoints = []; // construct endpoint by name
+                _.forEach(elements, function(element) {
+                    endpoints.push(element.Type + element.Id);
+                });
+                return endpoints;
+            }
+        });    
+    };
+      
+    /**
+     * @function composeSystem
+     * @description Compose a logical server using specified Elements
+     */
+    EmcComposeSystemJob.prototype.composeSystem = function() {
+        var self = this;
+        var parse = urlParse(self.redfish.settings.uri);
+        
+        if (0 > _.indexOf(['compose','recompose','destroy'], self.action)) {
+            throw new Error('Unknown Action Type: ' + self.action);
```

Comment:
Since `composeSystem` is used in the first promise chain (https://github.com/RackHD/on-tasks/pull/145/files#diff-cf2c2c498ad9eeacc1776e403cc45f85R67), so throwing error need be wrapped as a rejected promise, otherwise the `self._done(err)` will not be called as expected.